rotkehlchen/chain/ethereum/modules/weth/decoder.py

```diff
@@ -0,0 +1,159 @@
+
+    def _decode_deposit_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        depositor = hex_or_bytes_to_address(tx_log.topics[1])
+        deposited_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        deposited_amount = asset_normalized_value(amount=deposited_amount_raw, asset=self.eth)
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.counterparty in (WETH_CONTRACT, depositor) and
+                event.balance.amount == deposited_amount and
+                event.asset == self.eth
+            ):
+                if event.counterparty == depositor:
+                    return None, None
+
+                event.counterparty = CPT_WETH
+                event.event_type = HistoryEventType.DEPOSIT
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.notes = f'Wrap {deposited_amount} {self.eth.symbol} in {self.weth.symbol}'  # noqa: E501
+
+        in_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=self.base_tools.get_next_sequence_counter(),
+            timestamp=ts_sec_to_ms(transaction.timestamp),
+            location=Location.BLOCKCHAIN,
+            event_type=HistoryEventType.RECEIVE,
+            event_subtype=HistoryEventSubType.RECEIVE_WRAPPED,
+            asset=self.weth,
+            balance=Balance(amount=deposited_amount),
+            location_label=depositor,
+            counterparty=CPT_WETH,
+            notes=f'Receive {deposited_amount} {self.weth.symbol}',
+        )
+        return in_event, None
+
+    def _decode_withdrawal_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        withdrawer = hex_or_bytes_to_address(tx_log.topics[1])
+        withdrawn_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        withdrawn_amount = asset_normalized_value(amount=withdrawn_amount_raw, asset=self.eth)
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.counterparty in (WETH_CONTRACT, withdrawer) and
+                event.balance.amount == withdrawn_amount and
+                event.asset == self.eth
```

Comment:
I am confused. You seem to find the ETH receive event and to turn it into the WETH return event and then to create a new event for receiving ETH, essentially overwriting the original ETH receive event. Why?

Does the WETH contract have no normal transfer events?

Reply:
WETH contract has Transfer events but they're not emitted for deposit & withdrawal. Only when WETH is transferred

Reply:
Sure but in that case why change the order? You are essentially replacing one event with the other.

Reply:
Will change that. Misunderstood what you meant by **Unwrap at the Transfer event**. It's clear now though

Reply:
It's probably fine like this. But indeed I just realized that right now the receive event will appear probably before the send event which is gonna be weird ... :sweat_smile: so perhaps as you had before was better? It's just that the way it was written it was really confusing.

we generally try to have send event (out events) first and then receive (in events) since it (1) makes more sense and (2) is more readable in the ui